controller/src/controller/models/domain/rsptag.py

```diff
@@ -414,19 +526,50 @@ def _compare(self, other: object) -> int:
         if rank != 0:
             return rank
 
+        # If we have two tags with the same version, we should next check
+        # the RSP version.  If it doesn't exist, it should sort as a lower
+        # version than anything with an RSP version tag, to preserve backwards-
+        # compatibility.
+
+        rank = self._compare_rsp_versions(other)
+        if rank != 0:
+            return rank
+
         # semver ignores the build for sorting purposes, but we don't want to
         # since we want newer cycles to sort ahead of older cycles (and newer
         # cycle builds to sort above older cycle builds) in otherwise matching
         # tags, and the cycle information is stored in the build.
+        return self._compare_build_versions(other)
+
+    def _check_comparable_types(self, other: object) -> None:
+        if not isinstance(other, RSPImageTag):
+            raise NotImplementedError
+        if self.image_type != other.image_type:
+            raise NotImplementedError
+
+    def _compare_rsp_versions(self, other: Self) -> int:
+        if self.rsp_version is None or other.rsp_version is None:
+            if self.rsp_version is None:
+                if other.rsp_version is not None:
+                    return -1
+                return 0
+            if other.rsp_version is None:
+                return 1
+        return self.rsp_version.compare(other.rsp_version)
+
+    def _compare_build_versions(self, other: Self) -> int:
+        if self.version is None or other.version is None:
+            return NotImplemented
```

Comment:
Ugh, I don't like this. It implies that the version numbers might not be comparable, but the reality is that we already know that both versions have valid semvers. We've just lost the type information by passing in too-generic of parameters. I think the calling function needs to use local variables for the semvers so that you get type narrowing when you check that they're not `None`, and then this helper should take two semvers as parameters, which avoids the need for this check entirely.

Reply:
Fixed.